contracts/router/tree/TokenTree.sol

```diff
@@ -0,0 +1,578 @@
+
+    // ══════════════════════════════════════════════════ STORAGE ══════════════════════════════════════════════════════
+
+    // The nodes of the tree are stored in an array. The root node is at index 0.
+    Node[] internal _nodes;
+
+    // The list of all supported liquidity pools. All values are unique.
+    address[] internal _pools;
+
+    // (pool address => pool description)
+    mapping(address => Pool) internal _poolMap;
+
+    // (pool => token => tokenIndex) for each pool, stores the index of each token in the pool.
+    mapping(address => mapping(address => uint8)) public tokenIndexes;
+
+    // (token => nodes) for each token, stores the indexes of all nodes that represent this token.
+    mapping(address => uint256[]) internal _tokenNodes;
+
+    // The full path from every node to the root is stored using bitmasks in the following way:
+    // - For a node with index i at depth N, lowest N + 1 bytes of _rootPath[i] are used to store the path to the root.
+    // - The lowest byte is always the root index. This is always 0, but we store this for consistency.
+    // - The highest byte is always the node index. This is always i, but we store this for consistency.
+    // - The remaining bytes are indexes of the nodes on the path from the node to the root (from highest to lowest).
+    // This way the finding the lowest common ancestor of two nodes is reduced to finding the lowest differing byte
+    // in node's encoded root paths.
+    uint256[] internal _rootPath;
+
+    // (node => bitmask with all attached pools to the node).
+    // Note: This excludes the pool that connects the node to its parent.
+    mapping(uint256 => uint256) internal _attachedPools;
+
+    // ════════════════════════════════════════════════ CONSTRUCTOR ════════════════════════════════════════════════════
+
+    constructor(address bridgeToken) {
+        // The root node is always the bridge token
+        _addNode({token: bridgeToken, depth: 0, poolIndex: 0, rootPathParent: 0});
+        // Push the empty pool so that `poolIndex` for non-root nodes is never 0
+        _pools.push(address(0));
+    }
+
+    // ══════════════════════════════════════════════ INTERNAL LOGIC ═══════════════════════════════════════════════════
+
+    /// @dev Adds a pool with `N = tokensAmount` tokens to the tree by adding N-1 new nodes
+    /// as the children of the given node. Given node needs to represent a token from the pool.
+    function _addPool(
+        uint256 nodeIndex,
+        address pool,
+        address poolModule,
+        uint256 tokensAmount
+    ) internal {
+        require(nodeIndex < _nodes.length, "Out of range");
+        Node memory node = _nodes[nodeIndex];
+        if (poolModule == address(0)) poolModule = address(this);
+        (bool wasAdded, uint8 poolIndex) = (false, _poolMap[pool].index);
+        if (poolIndex == 0) {
+            poolIndex = uint8(_pools.length);
+            require(poolIndex <= type(uint8).max, "Too many pools");
```

Comment:
`poolIndex` should always fit in `uint8` given the unsafe cast on pools.length, so require check won't do the job here.

Issue is the unsafe cast which should produce duplicate pool indices. Instead, consider

```solidity
require(_pools.length <= type(uint8).max);
poolIndex = uint8(_pools.length);
```

Reply:
Would add a test for this require statement as well, since couldn't find in the PR

Reply:
Good find, thanks!
- Test was added in 322debf94c18fb61073fd082aa3b8ba9e2117756
- Fix was done in 91900feee2531d3dc6bedfb2bb6b46c7c0f30109